// mylittlepet/src/constants/vietnamese.js
// Vietnamese language constants for My Little Pet Admin
export const VIETNAMESE_TRANSLATIONS = {
    // Common
    common: {
        cancel: 'Hủy',
        save: 'Lưu',
        create: 'Tạo',
        update: 'Cập nhật',
        edit: 'Chỉnh sửa',
        delete: 'Xóa',
        add: 'Thêm',
        search: 'Tìm kiếm',
        actions: 'Hành động',
        status: 'Trạng thái',
        level: 'Cấp độ',
        name: 'Tên',
        type: 'Loại',
        price: 'Giá',
        quantity: 'Số lượng',
        description: 'Mô tả',
        stats: 'Thống kê',
        effects: 'Hiệu ứng',
        rarity: 'Độ hiếm',
        owner: 'Chủ sở hữu',
        showDetails: 'Hiển thị chi tiết',
        hideDetails: 'Ẩn chi tiết',
        viewDetails: 'Xem chi tiết',
        close: 'Đóng',
        loading: 'Đang tải...',
        noDataFound: 'Không tìm thấy dữ liệu',
        error: 'Lỗi',
        success: 'Thành công',
        warning: 'Cảnh báo',
        info: 'Thông tin',
        previous: 'Trước',
        next: 'Tiếp', page: 'Trang',
        of: 'của',
        showing: 'Hiển thị',
        to: 'đến',
        entries: 'mục',
        total: 'Tổng',
        all: 'Tất cả',
        yes: 'Có',
        no: 'Không',
        noPlayersFound: 'Không tìm thấy người chơi nào phù hợp với tiêu chí của bạn.'
    },    // Authentication
    auth: {
        login: 'Đăng nhập',
        logout: 'Đăng xuất',
        register: 'Đăng ký',
        email: 'Email',
        emailAddress: 'Địa chỉ Email',
        password: 'Mật khẩu',
        confirmPassword: 'Xác nhận mật khẩu',
        fullName: 'Họ và tên',
        enterEmail: 'Nhập email của bạn',
        enterPassword: 'Nhập mật khẩu của bạn',
        enterConfirmPassword: 'Nhập lại mật khẩu',
        enterFullName: 'Nhập họ và tên đầy đủ',
        signIn: 'Đăng nhập', signingIn: 'Đang đăng nhập...',
        signUp: 'Đăng ký',
        signingUp: 'Đang đăng ký...',
        signInToAdmin: 'Đăng nhập vào Hệ thống Quản trị',
        signUpForAccount: 'Đăng ký tài khoản quản trị viên',
        signOut: 'Đăng xuất',
        welcomeBack: 'Chào mừng trở lại',
        createAccount: 'Tạo tài khoản mới',
        signInToAccount: 'Đăng nhập vào tài khoản quản trị viên của bạn',
        rememberMe: 'Ghi nhớ đăng nhập',
        forgotPassword: 'Quên mật khẩu?',
        demoCredentials: 'Thông tin đăng nhập demo',
        confirmLogout: 'Xác nhận đăng xuất',
        confirmLogoutMessage: 'Bạn có chắc chắn muốn đăng xuất không?',
        staySignedIn: 'Tiếp tục đăng nhập',
        invalidCredentials: 'Email hoặc mật khẩu không hợp lệ',
        alreadyHaveAccount: 'Đã có tài khoản?',
        dontHaveAccount: 'Chưa có tài khoản?',
        goToLogin: 'Đăng nhập ngay',
        goToRegister: 'Đăng ký ngay',
        passwordsDoNotMatch: 'Mật khẩu không khớp',
        emailAlreadyExists: 'Email này đã được sử dụng',
        registrationSuccess: 'Đăng ký thành công! Bạn có thể đăng nhập ngay bây giờ.',
        registrationFailed: 'Đăng ký thất bại. Vui lòng thử lại.',
        weakPassword: 'Mật khẩu phải có ít nhất 8 ký tự, bao gồm chữ hoa, chữ thường và số'
    },

    // Navigation
    nav: {
        players: 'Người chơi',
        pets: 'Thú cưng',
        items: 'Vật phẩm',
        analytics: 'Phân tích',
        settings: 'Cài đặt',
        myLittlePet: 'Thú Cưng Bé Nhỏ Của Tôi',
        admin: 'Quản trị viên'
    },

    // Players
    players: {
        title: 'Quản Lý Người Chơi',
        management: 'Quản Lý Người Chơi',
        playerDetails: 'Chi Tiết Người Chơi',
        searchPlayers: 'Tìm kiếm người chơi...',
        allStatus: 'Tất Cả Trạng Thái',
        active: 'Hoạt động',
        banned: 'Bị cấm',
        inactive: 'Không hoạt động',
        player: 'Người chơi',
        pets: 'Thú cưng',
        items: 'Vật phẩm',
        achievements: 'Thành tích',
        registeredAt: 'Ngày đăng ký',
        lastActive: 'Hoạt động cuối',
        banPlayer: 'Cấm người chơi',
        unbanPlayer: 'Bỏ cấm người chơi',
        editPlayer: 'Chỉnh sửa người chơi',
        username: 'Tên người dùng',
        totalPets: 'Tổng Thú Cưng',
        totalItems: 'Tổng Vật Phẩm',
        totalAchievements: 'Tổng Thành Tích',
        registered: 'Đã đăng ký',
        inventory: 'Kho đồ',
        noPetsFound: 'Không tìm thấy thú cưng',
        noItemsFound: 'Không tìm thấy vật phẩm',
        noAchievementsFound: 'Không tìm thấy thành tích',
        unlocked: 'Đã mở khóa'
    },

    // Pets
    pets: {
        title: 'Quản Lý Thú Cưng',
        management: 'Quản Lý Thú Cưng',
        addPet: 'Thêm Thú Cưng',
        editPet: 'Chỉnh Sửa Thú Cưng',
        addNewPet: 'Thêm Thú Cưng Mới',
        searchPets: 'Tìm kiếm thú cưng...',
        allRarities: 'Tất Cả Độ Hiếm',
        allTypes: 'Tất Cả Loại',
        selectOwner: 'Chọn Chủ Sở Hữu',
        abilities: 'Kỹ năng', addAbility: 'Thêm',
        addAbilityPlaceholder: 'Thêm kỹ năng...',
        hp: 'HP',
        noPetsFound: 'Không tìm thấy thú cưng',
        noPetsFoundMessage: 'Không có thú cưng nào phù hợp với tiêu chí lọc hiện tại.', confirmDelete: 'Bạn có chắc chắn muốn xóa thú cưng này không?'
    },

    // Items
    items: {
        title: 'Quản Lý Vật Phẩm',
        management: 'Quản Lý Vật Phẩm',
        addItem: 'Thêm Vật Phẩm',
        editItem: 'Chỉnh Sửa Vật Phẩm',
        addNewItem: 'Thêm Vật Phẩm Mới',
        searchItems: 'Tìm kiếm vật phẩm...',
        allTypes: 'Tất Cả Loại',
        item: 'Vật phẩm',
        statName: 'Tên thuộc tính...',
        value: 'Giá trị...',
        totalItems: 'Số Vật Phẩm',
        totalQuantity: 'Tổng Số Lượng Vật Phẩm',
        totalValue: 'Tổng Giá Trị',
        noItemsFound: 'Không tìm thấy vật phẩm',
        noItemsFoundMessage: 'Không có vật phẩm nào phù hợp với tiêu chí lọc hiện tại.',
        confirmDelete: 'Bạn có chắc chắn muốn xóa vật phẩm này không?'
    },    // Item Types
    itemTypes: {
        food: 'Thức ăn',
        toy: 'Đồ chơi',
        medicine: 'Thuốc',
        accessory: 'Phụ kiện',
        consumable: 'Tiêu hao',
        material: 'Nguyên liệu'
    },

    // Pet Types
    petTypes: {
        dragon: 'Rồng',
        bird: 'Chim',
        beast: 'Thú',
        elemental: 'Nguyên Tố',
        fairy: 'Tiên',
        reptile: 'Bò Sát',
        magical: 'Phép Thuật',
        spirit: 'Linh Hồn',
        golem: 'Golem'
    },

    // Rarity
    rarities: {
        common: 'Thông Thường',
        uncommon: 'Không Phổ Biến',
        rare: 'Hiếm',
        epic: 'Sử thi',
        legendary: 'Huyền thoại',
        mythic: 'Thần thoại'
    },

    // Status
    statuses: {
        active: 'Hoạt động',
        inactive: 'Không hoạt động',
        banned: 'Bị cấm'
    },

    // Notifications
    notifications: {
        playerUpdated: 'Đã cập nhật thông tin người chơi thành công',
        playerBanned: 'Đã cấm người chơi thành công',
        playerUnbanned: 'Đã bỏ cấm người chơi thành công',
        playerDeleted: 'Đã xóa người chơi thành công',
        petCreated: 'Đã tạo thú cưng mới thành công',
        petUpdated: 'Đã cập nhật thú cưng thành công',
        petDeleted: 'Đã xóa thú cưng thành công',
        itemCreated: 'Đã tạo vật phẩm mới thành công',
        itemUpdated: 'Đã cập nhật vật phẩm thành công',
        itemDeleted: 'Đã xóa vật phẩm thành công',
        loginSuccess: 'Đăng nhập thành công',
        loginFailed: 'Đăng nhập thất bại',
        logoutSuccess: 'Đăng xuất thành công'
    },

    // Validation
    validation: {
        required: 'Trường này là bắt buộc',
        invalidEmail: 'Email không hợp lệ',
        passwordTooShort: 'Mật khẩu phải có ít nhất 6 ký tự',
        nameRequired: 'Tên là bắt buộc',
        levelRequired: 'Cấp độ là bắt buộc',
        priceRequired: 'Giá là bắt buộc',
        quantityRequired: 'Số lượng là bắt buộc',
        ownerRequired: 'Chủ sở hữu là bắt buộc'
    },

    // Error Messages
    errors: {
        generic: 'Đã xảy ra lỗi. Vui lòng thử lại.',
        networkError: 'Lỗi kết nối mạng. Vui lòng kiểm tra kết nối internet.',
        unauthorized: 'Bạn không có quyền truy cập.',
        notFound: 'Không tìm thấy tài nguyên.',
        serverError: 'Lỗi máy chủ. Vui lòng thử lại sau.',
        loadingFailed: 'Không thể tải dữ liệu'
    },

    // Buttons and Actions
    actions: {
        addNew: 'Thêm mới',
        saveChanges: 'Lưu thay đổi',
        discardChanges: 'Hủy thay đổi',
        confirmAction: 'Xác nhận',
        goBack: 'Quay lại',
        refresh: 'Làm mới',
        export: 'Xuất',
        import: 'Nhập',
        filter: 'Lọc',
        sort: 'Sắp xếp',
        resetFilter: 'Đặt lại bộ lọc'
    },

    // Pet Care Abilities
    abilities: {
        warmBreath: 'Hơi Thở Ấm Áp',
        protectiveRoar: 'Tiếng Gầm Bảo Vệ',
        healingScales: 'Vảy Rồng Chữa Lành',
        energyRecovery: 'Hồi Phục Năng Lượng',
        brightCrystalWings: 'Cánh Pha Lê Sáng',
        healingFire: 'Lửa Chữa Lành',
        gentleCaress: 'Vỗ Về Dịu Dàng',
        protectAllies: 'Bảo Vệ Đồng Đội',
        nightVision: 'Tầm Nhìn Ban Đêm',
        giftGems: 'Tặng Đá Quý',
        gentleShaking: 'Rung Lắc Nhẹ Nhàng',
        protectiveArmor: 'Giáp Bảo Vệ',
        coolWater: 'Tưới Nước Mát',
        healing: 'Hồi Máu',
        comfortingMist: 'Sương Mù Dễ Chịu',
        joyfulJump: 'Nhảy Vui Vẻ',
        cuteTickle: 'Cù Lét Dễ Thương',
        celestialDance: 'Nhảy Múa Thiên Thể',
        starHealing: 'Ánh Sao Chữa Lành',
        protectiveRoar2: 'Tiếng Gầm Bảo Vệ',
        gentleCooling: 'Làm Mát Dịu Dàng',
        protectiveIceShield: 'Khiên Băng Bảo Vệ',
        cuteCurl: 'Cuộn Tròn Dễ Thương',
        gracefulFlight: 'Bay Lượn Uyển Chuyển',
        melodiousSong: 'Hót Véo Von',
        gentleWingFlap: 'Cánh Đập Nhẹ Nhàng',
        healingAura: 'Hào Quang Hồi Máu',
        rainbow: 'Tia Cầu Vồng',
        purification: 'Thanh Tẩy',
        protectiveMagicShield: 'Khiên Phép Thuật Bảo Vệ',
        comfortableWarmth: 'Hơi Ấm Dễ Chịu',
        warmHeatWave: 'Sóng Nhiệt Ấm Áp',
        coldResistance: 'Kháng Lạnh',
        gentleStep: 'Bước Nhẹ Nhàng',
        connectionPortal: 'Cổng Kết Nối',
        warmEnergy: 'Năng Lượng Ấm Áp',
        joyfulTeleport: 'Dịch Chuyển Vui Vẻ'
    },

    // Time and Date
    time: {
        today: 'Hôm nay',
        yesterday: 'Hôm qua',
        thisWeek: 'Tuần này',
        lastWeek: 'Tuần trước',
        thisMonth: 'Tháng này',
        lastMonth: 'Tháng trước',
        thisYear: 'Năm này',
        lastYear: 'Năm trước',
        minutesAgo: 'phút trước',
        hoursAgo: 'giờ trước',
        daysAgo: 'ngày trước',
        weeksAgo: 'tuần trước',
        monthsAgo: 'tháng trước',
        yearsAgo: 'năm trước',
        justNow: 'Vừa xong'
    }
};

// Helper function to get translation
export function getTranslation(key) {
    const keys = key.split('.');
    let translation = VIETNAMESE_TRANSLATIONS;

    for (const k of keys) {
        if (translation && translation[k]) {
            translation = translation[k];
        } else {
            return key; // Return the key if translation not found
        }
    }

    return translation;
}

// Short alias for translation function
export const t = getTranslation;

export default VIETNAMESE_TRANSLATIONS;
